File: ACADEMIC CODES/Trie/python/trie_delete.py

def index(inputCh):
    return ord(inputCh) - ord('a')


class Node(object):
    children = None
    def __init__(self):
        self.children = [None] * 26
        self.ch = ' '
        self.word = ""
        self.isLeaf = False
        self.parent = None


class Trie(object):
    root = None

    def __init__(self):
        self.root = Node()



    def insertString(self, in_string):
        current = self.root
        for c in in_string[:-1]:
            idx = index(c)
            try:
                if current.children[idx] is None:
                    current.children[idx] = Node()
                    current.children[idx].ch = c
                    current.children[idx].word = current.word + c
                    current.children[idx].parent = current
            except IndexError as ie:
                print("idx = {}\nchar = {}\nstring = {}".format(idx, c, in_string))
            current = current.children[idx]
        current.isLeaf = True



    def searchString(self, in_string):
        current = self.root
        for c in in_string:
            idx = index(c)
            if current.children[idx] is not None:
                current = current.children[idx]
            else:
                return False
        return current.isLeaf


    def deleteString(self, in_string):
        current = self.root
        if self.searchString(in_string):
            for c in in_string:
                idx = index(c)
                current = current.children[idx]
            current.isLeaf = False


keys = ["cat\n","car\n","pick\n","pickle\n","dog\n","do\n"]
new_trie = Trie()
for key in keys:
    new_trie.insertString(key)


print(new_trie.searchString("pickle"))
print(new_trie.searchString("picke"))


new_trie.deleteString('pickle')


print("\nAfter Deletion: \n")
print(new_trie.searchString('pick'))   
print(new_trie.searchString('pickle'))  

File: ACADEMIC CODES/Trie/javascript/trie_delete.js
let Node = function () {
    this.endWord = false; 
    this.children = {}; 
    };
    
    let Trie = function () {
        this.root = new Node();
    
        this.insert = function (input) {
        let node = this.root;
        for (let i = 0; i < input.length; i++) {
            if (!node.children.hasOwnProperty(input[i])) {
            let temp = new Node();
            node.children[input[i]] = temp; 
            node = temp;
            } else {
            node = node.children[input[i]];
            }
        }
        node.endWord = true;
        // console.log(this.root);
        };
    
        this.isExists = (word) => {
        let node = this.root;
        for (let i = 0; i < word.length; i++) {
            if (node.children.hasOwnProperty(word[i])) {
            node = node.children[word[i]];
            } else {
            return false;
            }
        }
        return node.endWord;
        };
    
        this.delete = (word) => {
        let node = this.root;
        for (let i = 0; i < word.length; i++) {
            if (node.children.hasOwnProperty(word[i])) {
            node = node.children[word[i]];
            } else {
            return false;
            }
        }
        node.endWord = false;
        return true;
        };
    
        this.printAllWords = () => {
        let words = [];
        let node = this.root;
    
        let search = function (node, str) {
            for (const entry of Object.entries(node.children)) {
            let char = entry[0];
            node = entry[1];
    
            if (node.endWord) {
                words.push(str.concat(char));
            }
            search(node, str.concat(char));
            }
        };
        search(node, "");
        return words;
        };
    };
    
var trie = new Trie();

trie.insert("cat");
trie.insert("car");
trie.insert("pick");
trie.insert("pickle");
trie.insert("dog");
trie.insert("do");

console.log(trie.printAllWords());

trie.delete("pickle");
trie.delete("car");


console.log(trie.printAllWords());